### PR TITLE
feat(search): additional default filter

### PR DESCRIPTION
**Pull request by Rom1-B**
Replace https://github.com/glpi-project/glpi/pull/13977

Addition of default search filters, applied in addition to user filters. This filter can be disabled by the user.

image

Settings in dropdowns:

image

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | !24956

### CODE REVIEW

**Review by orthagh — CHANGES_REQUESTED** [review]
Another issue : when previewing results, the search is saved in SESSION.
I suggest to remove the entry  (in SESSION) after previewing end.

File: install/mysql/glpi-empty.sql [hunk 1]
```diff
@@ -9550,4 +9550,17 @@ CREATE TABLE `glpi_searches_criteriafilters` (
...
 
+DROP TABLE IF EXISTS `glpi_defaultfilters`;
+CREATE TABLE `glpi_defaultfilters` (
+  `id` int unsigned NOT NULL AUTO_INCREMENT,
+  `name` varchar(255) DEFAULT NULL,
+  `is_active` tinyint NOT NULL DEFAULT '1',
+  `comment` text DEFAULT NULL,
+  `itemtype` varchar(100) DEFAULT NULL,
+  PRIMARY KEY (`id`),
+  UNIQUE KEY `itemtype` (`itemtype`),
```
Comment (orthagh) [inline install/mysql/glpi-empty.sql:9561]: I want to discuss about this unicity key.
I'm not against it, but at least, a message should inform the admin when he tries to add a new filter when another one exists on the same itemtype.

Now, do we want to support only one filter per itemtype ?
Is there any technical issue having multiple ? (UI aside)

**Review by orthagh — COMMENTED** [review]
Before accepting, I'm still waiting for alert feature when unicity error triggers

**Review by orthagh — APPROVED** [review]

**Review by trasher — APPROVED** [review]

File: src/DefaultFilter.php [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+            [
+                'name'  => 'itemtype',
+                'label' => __('Itemtype'),
+                'type'           => 'itemtypename',
+                'itemtype_list'      => 'globalsearch_types',
+            ]
+        ];
+    }
+
+    public static function getSearchCriteria($itemtype)
```
Comment (cedric-anne) [inline src/DefaultFilter.php:116]:
```suggestion
    public static function getSearchCriteria(string $itemtype): ?array
```

File: src/DefaultFilter.php [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+                'id' => $item['id'],
+                'name' => $item['name'],
+                'comment' => $item['comment'],
+                'search_criteria' => [
+                    'link' => 'AND',
+                    'criteria' => json_decode($item['search_criteria'], true),
+                ]
+            ];
+        }
+        return [];
```
Comment (cedric-anne) [inline src/DefaultFilter.php:163]: It should return `null` when there is nothing to return.
```suggestion
        return null;
```

File: templates/components/search/controls.html.twig [hunk 4]
```diff
@@ -222,6 +222,19 @@ $(document).ready(function() {
...
+      var url = window.location.href;
+      url = url.replace(/&?nodefault/g, '');
+      if (!$(this).is(":checked")) {
+         if (url.indexOf('?') == -1)
+            url += '?nodefault';
+         else
+            url += '&nodefault';
+      }
+      url = url.replace(/\?$/g, '');
+      window.location.replace(url);
```
Comment (cedric-anne) [inline templates/components/search/controls.html.twig:235]: Using regex does not seems safe. Using something like the following code would be more resilient to changes in URLs:
```suggestion
      const search_params = new URLSearchParams(window.location.search);
      if (!$(this).is(":checked")) {
          search_params.set('nodefault', '1');
      } else {
          search_params.delete('nodefault');
      }
      window.location.replace('?' + search_params.toString());
```

**Review by cedric-anne — CHANGES_REQUESTED** [review]

**Review by cedric-anne — COMMENTED** [review]
When default filter is enabled, the default criterion seems to be added, additionaly to the current search criteria. It may result in a duplicated default criterion:
image

I think this default criterion should not be added when default filters are enabled.

**Review by cedric-anne — APPROVED** [review]

**Pull request merged**